### PR TITLE
Add IBM WebSphere MQ Channel Name Bruteforce auxiliary module

### PR DESCRIPTION
**Pull request by pkb1s**
Add IBM WebSphere MQ Channel Name Bruteforce auxiliary module

Uses a dictionary to bruteforce MQ channel names. For all identified channels it also returns if SSL is used and whether it is a server-connection channel.

* IBM Downloads page: https://developer.ibm.com/messaging/mq-downloads/
* Tested on IBM MQ 7.5, 8 and 9
* Usage:
  * Download and install MQ Server
  * Create a new Queue Manager
  * Create a new channel (without SSL)
  * Run the module with a valid channel name. This can be obtained from the "ibm_channel_brute" module

## Example Output
``` 
msf > use auxiliary/scanner/misc/ibm_mq_channel_brute
msf auxiliary(ibm_mq_channel_brute) > set channels_file channel_list.txt
channels_file => channel_list.txt
msf auxiliary(ibm_mq_channel_brute) > set rhosts 10.1.1.144
rhosts => 10.1.1.144
msf auxiliary(ibm_mq_channel_brute) > set rport 1414
rport => 1414
msf auxiliary(ibm_mq_channel_brute) > run

[*] 10.1.1.144:1414       - Found channel: TEST.CHANNEL, IsEncrypted: False, IsMQI: True

[+] 10.1.1.144:1414       - Channels found: ["TEST.CHANNEL"]
[+] 10.1.1.144:1414       - Unencrypted MQI Channels found: ["TEST.CHANNEL"]

[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf auxiliary(ibm_mq_channel_brute) > 

```

## Example Verbose Output
```
msf auxiliary(ibm_mq_channel_brute) > set verbose true 
verbose => true
msf auxiliary(ibm_mq_channel_brute) > run

[*] 10.1.1.144:1414       - 10.1.1.144:1414 - Sending request for TEST.CHANNEL...
[*] 10.1.1.144:1414       - 10.1.1.144:1414 - Sending request for ADMIN.CHANNEL...
[*] 10.1.1.144:1414       - 10.1.1.144:1414 - Sending request for FALSE.CHANNEL...
[*] 10.1.1.144:1414       - 10.1.1.144:1414 - Sending request for PRIV.CHANNEL...
[*] 10.1.1.144:1414       - Found channel: TEST.CHANNEL, IsEncrypted: False, IsMQI: True

[+] 10.1.1.144:1414       - Channels found: ["TEST.CHANNEL"]
[+] 10.1.1.144:1414       - Unencrypted MQI Channels found: ["TEST.CHANNEL"]

[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf auxiliary(ibm_mq_channel_brute) > 
```

### CODE REVIEW

File: modules/auxiliary/scanner/misc/ibm_mq_channel_brute.rb [hunk 1]
```diff
@@ -0,0 +1,169 @@
+##
+# This module requires Metasploit: http://metasploit.com/download
```
Comment (bcoles) [inline modules/auxiliary/scanner/misc/ibm_mq_channel_brute.rb]:
```suggestion
# This module requires Metasploit: https://metasploit.com/download
```

File: modules/auxiliary/scanner/misc/ibm_mq_channel_brute.rb [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+      'Name'        => 'IBM WebSphere MQ Channel Name Bruteforce',
+      'Description' => 'This module uses a dictionary to bruteforce MQ channel names. For all identified channels it also returns if SSL is used and whether it is a server-connection channel.',
+      'Author'      => 'Petros Koutroumpis',
+      'License'     => MSF_LICENSE
+    )
+    register_options([
+      Opt::RPORT(1414),
+      OptInt.new('TIMEOUT', [true, "The socket connect timeout in seconds", 5]),
+      OptInt.new('CONCURRENCY', [true, "The number of concurrent channel names to check", 10]),
+      OptString.new('CHANNELS_FILE',
```
Comment (bcoles) [inline modules/auxiliary/scanner/misc/ibm_mq_channel_brute.rb]:
```suggestion
      OptPath.new('CHANNELS_FILE',
```

File: modules/auxiliary/scanner/misc/ibm_mq_channel_brute.rb [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+  def run_host(ip)
+    @channels = []
+    @unencrypted_mqi_channels = []
+    begin
+      channel_list
+      rescue ::Rex::ConnectionError
+      rescue ::Exception => e
+        print_error("#{e} #{e.backtrace}")
+      end
+      report_service(:host => rhost, :port => rport, :name => "finger")
```
Comment (bcoles) [inline modules/auxiliary/scanner/misc/ibm_mq_channel_brute.rb]: `"finger"` ?

File: modules/auxiliary/scanner/misc/ibm_mq_channel_brute.rb [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+      end
+    end
+  end
+
+  def get_channel_names
+    if(! @common)
+      File.open(datastore['CHANNELS_FILE'], "rb") do |fd|
+        data = fd.read(fd.stat.size)
+        @common = data.split(/\n/).compact.uniq
+        @common.delete("")
```
Comment (bcoles) [inline modules/auxiliary/scanner/misc/ibm_mq_channel_brute.rb]: Pretty sure this line does nothing.

```suggestion
```

File: modules/auxiliary/scanner/misc/ibm_mq_channel_brute.rb [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+    @channels = []
+    @unencrypted_mqi_channels = []
+    begin
+      channel_list
+      rescue ::Rex::ConnectionError
+      rescue ::Exception => e
+        print_error("#{e} #{e.backtrace}")
+      end
+      report_service(:host => rhost, :port => rport, :name => "finger")
+      print("\n")
```
Comment (bcoles) [inline modules/auxiliary/scanner/misc/ibm_mq_channel_brute.rb]:
```suggestion
      print_line
```

File: modules/auxiliary/scanner/misc/ibm_mq_channel_brute.rb [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+        print_status("#{ip}:#{rport} No channels found.")
+      else
+        print_good("Channels found: #{@channels}")
+        print_good("Unencrypted MQI Channels found: #{@unencrypted_mqi_channels}")
+        report_note(
+          :host => rhost,
+          :port => rport,
+          :type => 'mq.channels'
+        )
+      print("\n")
```
Comment (bcoles) [inline modules/auxiliary/scanner/misc/ibm_mq_channel_brute.rb]:
```suggestion
      print_line
```

**Comment by asoto-r7** [comment]
## Release Notes

`auxiliary/scanner/misc/ibm_mq_channel_brute` brute forces a list of channel names in IBM WebSphere MQ.

**Comment by asoto-r7** [comment]
## Release Notes

Given a queue manager and a channel, `auxiliary/scanner/misc/ibm_mq_login` brute forces credentials in IBM WebSphere MQ.

**Pull request merged**